### DISCUSSION TITLE
can I import my original flash save files (.sol) into ruffle?

### DISCUSSION THREAD

**Discussion by nicomicap** [ruffle-rs/ruffle#3816] [Q&A]
I want to continue right where I left off in a game but when I load it up with ruffle (on firefox) it shows as if i've never played it

**Comment by n0samu** [comment]
Ruffle now has a save manager (added in #10040) allowing you to replace a game's save data with an existing save file. Just right-click the game and click "Open Save Manager".

**Comment by Croworbit** [comment]
here is save locations should anyone in the future find this discussion when looking for save data, but this is from the discord faq (of course discord is a terrible place to put information that everyone needs, but eh)

this is for desktop
```
Linux:   /home/<username>/.local/share/ruffle

macOS: /Users/<username>/Library/Application Support/ruffle

Note: Depending on the level of trust your device attributes to the download, this can be the physical directory or the virtual/sandboxed version of it.
example:
/Users/<username>/Library/Containers/rs.ruffle.ruffle/Data/Library/Application\ Support/ruffle/`


Windows: C:\Users\<username>\AppData\Local\Ruffle

In this directory, you will find log files, a preferences.toml  configuration file if you have any settings customized, as well as shared object library (save) data. ```

**Comment by orazioedoardo** [comment]
I'm not seeing "Open Save Manager" in the context menu, on the web. What should I do to enable it?

**Reply by Croworbit** [reply]
if you don't see it, most likely no save data has been created yet. (on web)
the game may not save at all, or you may not have reached a point in the game where it does save

some games save after the game is closed, some save at points you can interact, some auto save and some do not save at all, even on original flash

**Reply by orazioedoardo** [reply]
Thanks indeed at some point the game saved some data as after closing and returning to the game, the progress was kept. However I tried to restore by uploading the .sol file it and it didn't work. Moreover I had a game that saves new data with the index of the level appended, and when trying to restore you can only replace a file, not upload a new one.

**Reply by Croworbit** [reply]
in the case of reuploading and it not working, if it is from a different version of the game than before, it may not work, (ie version number of the game)if the author changed save style (new internal format the dev created for their .sol) or if the game looks for filenames

(ie, if the game embedded in a site calls itself gamezed.swf , if you go to a different website where the same exact game is called 62345_gamezed.swf, they both create saves that the other will ignore)

it also might be a real issue, so you can open a github issue and upload the game and save data 

for the case of only being allowed to replace save data, keep your old data, allow a new save to get created, this makes a save to be replaced. replace it with your old data

I'm not sure about the part where that game saves data by index of level, so I'm not sure, might also be worth an issue

**Reply by Croworbit** [reply]
it might also be the case where since the game only saves when closed, you replacing data will have no effect, as once closed it will overwrite your replaced data with the queued data

and if you can only replace the data if the game is open, replacing does not work as intended. 
that would also be worth reporting, if that is the case

**Reply by orazioedoardo** [reply]
> in the case of reuploading and it not working, if it is from a different version of the game than before

It's the sol file I downloaded moments before from the save manager. Then I clear site data, open the game again, upload the sol, but progress was not preserved.

> for the case of only being allowed to replace save data, keep your old data, allow a new save to get created, this makes a save to be replaced. 

Another game I tried has several levels. If your character dies, you go back to the main menu where you can select one of the levels you have reached before dying. The save data for the level seems to be named game_lv1, game_lv2 and so on. When starting fresh these save data are not available thus cannot be replaced in the UI. Obviously I can play the game and unlock the levels again but then I don't need the save files.

**Reply by Croworbit** [reply]
`Obviously I can play the game and unlock the levels again but then I don't need the save files.`

ah, I had feared that would be the case for some games. 
this may *also* be worth mentioning in an issue/feature request

I'm of the opinion that ruffle's web save manager isn't ideal and it seems the problems you are having help prove that

**Reply by orazioedoardo** [reply]
> it might also be the case where since the game only saves when closed, you replacing data will have no effect, as once closed it will overwrite your replaced data with the queued data

What constitutes a closing the game though? I'm playing in a web UI. Is the game able to detect that the tab is closing and save the game?

What happens I think is that you can only replace save data after the game has been started and saved something named after what you want to replace, at which point the game doesn't read the save data again.

**Reply by Croworbit** [reply]
https://github.com/ruffle-rs/ruffle/issues/18544
is my issue on the save manager's issues (and other save data in flash game gripes)

maybe you can leave a comment there about what you think, or (better idea, really) open a new issue discussing your problems, (better chance of help)

**Reply by orazioedoardo** [reply]
Also the save manager seems to be "global", it includes every save data in browser storage for every game. Probably it should be scoped per-some-identifier of the SWF.

**Reply by Croworbit** [reply]
> what constitutes a game close

also a good question, my guess, page reload
